### ISSUE TITLE
com.sun.ts.tests.jsp.spec.el.jsp.ELJspVersionTag relies on implementation detail

### ISSUE THREAD

**Issue by brideck** [eclipse-ee4j/jakartaee-tck#535]
com.sun.ts.tests.jsp.spec.el.jsp.ELJspVersionTag was updated for EE9 as part of PR #349 with:

`JspWriter out = getJspContext().getOut();`
`out.println(poundExpr);`

And expects a result of "ValueExpression[#{foo}]".  But this behavior is relying on the RI's [overridden toString method](https://github.com/eclipse-ee4j/el-ri/blob/738dd355879d88f09269a53ab0418d4b5f9ee46d/impl/src/main/java/com/sun/el/ValueExpressionImpl.java#L260).

In other implementations, where this toString method is not overridden, the out.println statement simply results in output like "org.apache.jasper.el.JspValueExpression@20562".  This can be seen in testing with both Tomcat and Open Liberty.

A better solution here would be to use something in the API for validation, like:
`out.println(poundExpr.getExpressionString());`

This would result in the old output of "#{foo}" if everything is working properly.

**Comment by scottmarlow** [comment]
Thanks Brian!

Just sent email to https://www.eclipse.org/lists/jsp-dev/2020/Sep/index.html asking for input from Jakarta Server Pages team regarding this issue and suggested change.  The mail didn't show up in archives yet but should soon.

**Comment by scottmarlow** [comment]
CC @gurunrao

**Comment by scottmarlow** [comment]
I do agree that org.apache.jasper.el.JspValueExpression shouldn't be expected to implement toString method.  

Regarding the suggestion to call `poundExpr.getExpressionString())`, does that fulfill the test strategy and assertions for this JSP test?  

Javadoc from https://jakarta.ee/specifications/expression-language/4.0/apidocs/jakarta/el/Expression.html#getExpressionString--:
```
Returns the original String used to create this Expression, unmodified.

This is used for debugging purposes but also for the purposes of comparison (e.g. to ensure the expression in a configuration file has not changed).

This method does not provide sufficient information to re-create an expression. Two different expressions can have exactly the same expression string but different function mappings. Serialization should be used to save and restore the state of an Expression. 
```

**Comment by gurunrao** [comment]
@scottmarlow @brideck 
 I have made changes here: https://github.com/eclipse-ee4j/jakartaee-tck/compare/master...gurunrao:jsp-fix
CI runs are passing here:  https://ci.eclipse.org/jakartaee-tck/job/jakartaee-tck-guru/job/jsp-fix/3/testReport/
Let me know your inputs.

**Comment by brideck** [comment]
@gurunrao Those changes look sound to me.  Thanks for taking a look.

**Comment by scottmarlow** [comment]
Thanks @gurunrao , I did loop the jsp-dev@eclipse.org community in via an email discussion.  I asked them to give their feedback on the Jakarta Server Pages test changes made so far for ElJspVersionTag and will update that discussion to include `https://github.com/eclipse-ee4j/jakartaee-tck/compare/master...gurunrao:jsp-fix`.

I will mention that on Monday we will merge the `https://github.com/eclipse-ee4j/jakartaee-tck/compare/master...gurunrao:jsp-fix` change if we don't here feedback from them.

**Comment by markt-asf** [comment]
Looking at this now...

**Comment by markt-asf** [comment]
The changes made to the JSP TCK in #349 are definitely incorrect. I'm now reviewing the fix proposed by @gurunrao to ensure the test still meets the original aims.

**Comment by markt-asf** [comment]
OK. I think I have got to the bottom of what is going on.
In 3ec291f the `el_jsp_2.0.tld` was incorrectly updated to 3.0. The test requires that it uses a version less than 2.1. A related, but less important issue, is that 26286ba updated the schema location but not the version for `el_jsp.tld` resulting in an invalid TLD file.
The changes in 3ec291f broke the test (confirmed by building TCK from source before and after 3ec291f and testing with Tomcat).
The changes in #349 were an attempt to fix the test. The #349 fix addressed symptoms rather than the root cause and did so in a way that was both Glassfish specific and further broke the test.
@gurunrao's proposed changes address the Glassfish specific issues with the changes in #349 but do not address the issues introduced in 3ec291f nor the non-Glassfish specific issues introduced in #349.
I believe that the correct solution is to revert the JSP TCK changes introduced in #349 and to switch `el_jsp_2.0.tld` to the TLD 2.0 schema. I'm testing this now and if successful will provide a PR. I'll also fix the schema definition for `el_jsp.tld`

**Comment by markt-asf** [comment]
Confirmed that fixes the broken test. PR to follow once I've run the complete JSP TCK to check for additional issues.

**Comment by markt-asf** [comment]
Pull request to fix this issue available at #536
